File: apps/cargo-tracker/app/cargotracker/tracking/domain/model/entities/TrackingExceptionEvent.scala
package cargotracker.tracking.domain.model.entities

import cargotracker.tracking.domain.model.enums.ExceptionType
import cargotracker.tracking.domain.model.valueobjects.{TrackingExceptionEventId, TrackingLocation}

import java.time.Instant

/** 追跡例外イベント（US19 遅延 / US20 破損・紛失 / IT7、IT8 0.5 で id 追加）。
  *
  *   - `id`: 永続化 PK (`TrackingExceptionEventId`)。未保存時は None、保存後は Some。 IT8 0.5 (H5) で複合キー UPDATE を PK 直接更新に切り替えるため導入
  *   - `exceptionType`: 例外種別 (Delay / Damage / Lost / CustomsHold)
  *   - `location`: 発生場所
  *   - `occurredAt`: 発生時刻
  *   - `description`: 詳細記述（任意）
  *   - `escalationFlag`: 管理職エスカレーション要否 (Lost は強制 true)
  *   - `resolvedAt`: 対応完了時刻 (None = 未対応)
  *   - `resolutionNotes`: 対応報告内容
  */
final case class TrackingExceptionEvent(
    exceptionType: ExceptionType,
    location: TrackingLocation,
    occurredAt: Instant,
    description: Option[String] = None,
    escalationFlag: Boolean = false,
    resolvedAt: Option[Instant] = None,
    resolutionNotes: Option[String] = None,
    id: Option[TrackingExceptionEventId] = None
):
  def isResolved: Boolean = resolvedAt.isDefined
  def isActive: Boolean = !isResolved

File: apps/cargo-tracker/src/CargoTracker.Web/Tracking/Domain/Model/TrackingExceptionEvent.cs
namespace CargoTracker.Tracking.Domain.Model;

/// <summary>
/// 追跡例外イベント（集約内エンティティ・US19/US20）。遅延・破損・紛失などの例外事象を時系列で記録する。
/// 紛失（Lost）は登録時にエスカレーションフラグが立ち、上位管理者へのエスカレーション対象となる（domain-model ビジネスルール 3）。
/// </summary>
public sealed class TrackingExceptionEvent
{
    public ExceptionType ExceptionType { get; }
    public TrackingLocation Location { get; }
    public DateTimeOffset OccurredAt { get; }
    public string? Description { get; }
    public bool EscalationFlag { get; }
    public DateTimeOffset? ResolvedAt { get; private set; }
    public string? ResolutionNotes { get; private set; }

    public TrackingExceptionEvent(
        ExceptionType exceptionType, TrackingLocation location,
        DateTimeOffset occurredAt, string? description,
        DateTimeOffset? resolvedAt = null, string? resolutionNotes = null)
    {
        ArgumentNullException.ThrowIfNull(location);
        ExceptionType = exceptionType;
        Location = location;
        OccurredAt = occurredAt;
        Description = description;
        // 紛失は重大例外のためエスカレーション必須（domain-model BR3）。
        EscalationFlag = exceptionType == ExceptionType.Lost;
        ResolvedAt = resolvedAt;
        ResolutionNotes = resolutionNotes;
    }

    /// <summary>未解決（対応中）かどうか。</summary>
    public bool IsActive => ResolvedAt is null;

    /// <summary>例外を解決済みにする（US19/US20 の対応報告）。</summary>
    public void Resolve(DateTimeOffset resolvedAt, string? resolutionNotes)
    {
        ResolvedAt = resolvedAt;
        ResolutionNotes = resolutionNotes;
    }
}
